src/datarobotx/idp/custom_application_source.py

```diff
@@ -0,0 +1,320 @@
+        if custom_application_source["name"] == name:
+            return str(custom_application_source["id"])
+
+    create_custom_application_source_response = client.post(
+        "customApplicationSources/",
+    )
+    custom_application_source_id = create_custom_application_source_response.json()["id"]
+
+    client.patch(
+        f"customApplicationSources/{custom_application_source_id}",
+        json={"name": name},
+    )
+    return str(custom_application_source_id)
+
+
+# Helper functions
+def _find_existing_custom_application_source_version(
+    endpoint: str, token: str, custom_application_source_id: str, version_token: str
+) -> str:
+    client = dr.Client(endpoint=endpoint, token=token)  # type: ignore
+    versions = client.get(
+        f"customApplicationSources/{custom_application_source_id}/versions/"
+    ).json()["data"]
+    for version in versions:
+        if version.get("label") and version_token in version["label"]:
+            return str(version["id"])
+    raise KeyError("No matching application source version found")
+
+
+def _get_or_create_custom_application_source_version(
+    from_previous: bool,
+    endpoint: str,
+    token: str,
+    custom_application_source_id: str,
+    previous_custom_application_source_version_id: Optional[str] = None,
+    runtime_parameter_values: Optional[List[Union[RuntimeParameterValue, Dict[str, str]]]] = None,
+    **kwargs: Any,
+) -> str:
+    client = dr.Client(token=token, endpoint=endpoint)  # type: ignore
+
+    if runtime_parameter_values is not None:
+        runtime_parameter_values_objs = [
+            RuntimeParameterValue(**param) if isinstance(param, dict) else param
+            for param in runtime_parameter_values
+        ]
+    else:
+        runtime_parameter_values_objs = []
+
+    folder_path = kwargs.pop("folder_path", None)
+
+    if from_previous and not previous_custom_application_source_version_id:
+        raise ValueError("Must provide a previous version ID when creating from previous version.")
+
+    if from_previous:
+        # Get the previous version's details and use them as the base for the new version
+        previous_version = client.get(
+            f"customApplicationSources/{custom_application_source_id}/versions/{previous_custom_application_source_version_id}/"
+        ).json()
+        version_token = previous_version["label"].split("\nChecksum: ")[-1]
+    else:
+        version_token = get_hash(
+            Path(folder_path) if folder_path else None,
+            custom_application_source_id,
+            runtime_parameter_values=runtime_parameter_values_objs,
+            **kwargs,
+        )
+
+    try:
+        existing_version_id = _find_existing_custom_application_source_version(
+            endpoint=endpoint,
+            token=token,
+            custom_application_source_id=custom_application_source_id,
+            version_token=version_token,
+        )
+        return existing_version_id
```

Comment:
Actually reading this again, I think there is a logical problem: 

Lets say user passes `from_previous` to this function. But their runtime parameters are different. You create a token from the version already on DataRobot but that token doesn't involve the runtime parameters. When you try to return the `existing_version_id`, it succeeds and the runtime parameters are therefore never updated

Reply:
right. agree